### PR TITLE
[FCOS] Fix Terraform for Azure

### PR DESCRIPTION
**Pull request by jomeier**
Set blob type for ignition file for Bootstrap VM to 'Block' instead of 'Page'.

Change source_uri to source because on OKD we download the FCOS image locally before uploading it to Azure.

### CODE REVIEW

**Comment by walidshaari** [comment]
in the last installer https://origin-release.apps.ci.l2s4.p1.openshiftapps.com/releasestream/4-stable/release/4.5.0-0.okd-2020-08-12-020541 blob type for ignition file for Bootstrap VM to 'Block' instead of 'Page'.

one more file needs changing  **pkg/tfvars/azure/azure.go b/pkg/tfvars/azure/azure.go**, to include back @jomeier old code that was dropped out after Beta5
```
        "github.com/Azure/go-autorest/autorest/to"
+        "github.com/pkg/errors"
 
+       "github.com/openshift/installer/pkg/tfvars/internal/cache"
        "github.com/openshift/installer/pkg/types"
        "github.com/openshift/installer/pkg/types/azure/defaults"
        azureprovider "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
@@ -67,6 +69,11 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
                emulateSingleStackIPv6 = true
        }
 
+        cachedImage, err := cache.DownloadImageFile(sources.ImageURL)
+       if err != nil {
+               return nil, errors.Wrap(err, "failed to use cached Azure image")
+       }
+
```

**Comment by jomeier** [comment]
> in the last installer, https://origin-release.apps.ci.l2s4.p1.openshiftapps.com/releasestream/4-stable/release/4.5.0-0.okd-2020-08-12-020541 blob type for ignition file for Bootstrap VM to 'Block' instead of 'Page' was already updated, however not the source_uri and not the below cache file code.
> 
> one more file needs changing **pkg/tfvars/azure/azure.go b/pkg/tfvars/azure/azure.go**, to include back @jomeier old code that was dropped out after Beta5
> 
> ```
>         "github.com/Azure/go-autorest/autorest/to"
> +        "github.com/pkg/errors"
>  
> +       "github.com/openshift/installer/pkg/tfvars/internal/cache"
>         "github.com/openshift/installer/pkg/types"
>         "github.com/openshift/installer/pkg/types/azure/defaults"
>         azureprovider "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
> @@ -67,6 +69,11 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
>                 emulateSingleStackIPv6 = true
>         }
>  
> +        cachedImage, err := cache.DownloadImageFile(sources.ImageURL)
> +       if err != nil {
> +               return nil, errors.Wrap(err, "failed to use cached Azure image")
> +       }
> +
> ```

@walid: The code in /pkg/tfvars/azure/azure.go is ok in latest fcos branch.

File: data/data/azure/main.tf [hunk 1]
```diff
@@ -170,8 +170,8 @@ resource "azurerm_storage_blob" "rhcos_image" {
   storage_account_name   = azurerm_storage_account.cluster.name
   storage_container_name = azurerm_storage_container.vhd.name
   type                   = "Page"
-  source_uri             = var.azure_image_url
-  metadata               = map("source_uri", var.azure_image_url)
+  source                 = var.azure_image_url
+  metadata               = map("source", var.azure_image_url)
```
Comment (LorbusChris) [inline data/data/azure/main.tf:174]: why is this change necessary?

File: data/data/azure/main.tf [hunk 2]
```diff
@@ -170,8 +170,8 @@ resource "azurerm_storage_blob" "rhcos_image" {
   storage_account_name   = azurerm_storage_account.cluster.name
   storage_container_name = azurerm_storage_container.vhd.name
   type                   = "Page"
```
Comment (LorbusChris) [inline data/data/azure/main.tf:172]: Page is ok here?

Comment (jomeier) [reply data/data/azure/main.tf:172]: Yes. But only for VHD files.

My initial commit was overwritten by someone force pushing to the installer repo. 

Vadim restored my commit partially. Because of that "Page" is already in the fcos branch. But source_uri is missing.

In addition someone was setting Page also to the bootstrap.ign which does require proper byte alignment that bootstrap.ign does not have, throwing an error. And also this is not required.

I know: confusing.

**Comment by LorbusChris** [comment]
@jomeier ah yeah the original change was from Block to Page. Why does the bootstrap VM need the Block type now?

**Comment by jomeier** [comment]
> @jomeier ah yeah the original change was from Block to Page. Why does the bootstrap VM need the Block type now?

It's not the bootstrap VM but the bootstrap.ign file that is uploaded to the blob store. I haven't set it to 'Page' before. If you set it to Page, Terraform / the Azure API complains about that the bootstrap.ign is not aligned to 512 Byte boundaries ...

In the master branch it's set to "Block". This is correct.

**Comment by LorbusChris** [comment]
Right. This error was introduced with https://github.com/openshift/installer/commit/35a07c76ff185be4e9d5ef2a2b112a998f324b37#diff-334bba265e5f181be06ef444a82e78c9. It should never have become Page there.
/approve
/lgtm

**Comment by arunvel1988** [comment]
Your documentation really sucks.
OKD on azure not working.
ERROR Error: Error creating Blob "rhcoswii3l.vhd" (Container "vhd" / Account "clusterwii3l"): Error opening source file for upload "https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/34.20210529.3.0/x86_64/fedora-coreos-34.20210529.3.0-azure.x86_64.vhd.xz": open https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/34.20210529.3.0/x86_64/fedora-coreos-34.20210529.3.0-azure.x86_64.vhd.xz: no such file or directory
Facing above error. Poor documentation
